src/isolation/landlock.rs

```diff
@@ -0,0 +1,212 @@
+
+		UhyveLandlockWrapper {
+			rw_paths: rw_paths.to_vec(),
+			ro_paths: ro_paths.to_vec(),
+			ro_dirs: ro_dirs.to_vec(),
+			compat_level: Self::determine_compat_level(sandbox_mode),
+		}
+	}
+
+	pub fn apply_landlock_restrictions(&self) {
+		{
+			let _status = match Self::enforce_landlock(self) {
+				Ok(status) => status,
+				Err(error) => panic!("Unable to initialize Landlock: {error:?}"),
+			};
+		}
+	}
+
+	/// Using the file sandbox mode provided by the user, we derive the corresponding
+	/// Landlock compatibility mode. This deliberately omits the CompatLevel::SoftRequirement
+	/// possibility, while leaving room for future enhancements to the host file access
+	/// layer
+	///
+	/// `sandbox_mode` - User-provided `FileSandboxMode`
+	fn determine_compat_level(sandbox_mode: FileSandboxMode) -> CompatLevel {
+		match sandbox_mode {
+			FileSandboxMode::None => panic!(),
+			FileSandboxMode::BestEffort => CompatLevel::BestEffort,
+			FileSandboxMode::Strict => CompatLevel::HardRequirement,
+		}
+	}
+
+	/// Initializes Landlock by providing R/W-access to user-defined and
+	/// Uhyve-defined paths.
+	fn enforce_landlock(&self) -> Result<RestrictionStatus, RestrictError> {
+		// This should be incremented regularly.
+		let abi = ABI::V5;
+		// Used for explicitly whitelisted files (read & write).
+		let access_all: landlock::BitFlags<AccessFs, u64> = AccessFs::from_all(abi);
+		// Used for the kernel itself, as well as "system directories" that we only read from.
+		let access_read: landlock::BitFlags<AccessFs, u64> = AccessFs::from_read(abi);
+
+		// LANDLOCK_ACCESS_FS_REMOVE_FILE and LANDLOCK_ACCESS_FS_REMOVE_DIR do not apply to
+		// whitelisted files themselves, but apply to the contents of directories transitively.
+		// This means that we have to give Uhyve read-only access to the parent directory with
+		// some additional permissions so as to support fs::remove_file and removing the temporary
+		// directory.
+		let access_dir_with_rm: landlock::BitFlags<AccessFs, u64> =
+			AccessFs::ReadDir | AccessFs::RemoveDir | AccessFs::RemoveFile;
+
+		debug!(
+			"Enabling Landlock path isolation with following compatibility mode: {:?}",
+			self.compat_level
+		);
+
+		Ok(Ruleset::default()
+			.handle_access(access_all)?
+			.create()?
+			.add_rules(
+				self.rw_paths
+					.as_slice()
+					.iter()
+					.map::<Result<_, RestrictError>, _>(|p| {
+						Ok(PathBeneath::new(PathFd::new(p)?, AccessFs::from_all(abi)))
+					}),
+			)?
+			.add_rules(
+				self.ro_paths
+					.as_slice()
+					.iter()
+					.map::<Result<_, RestrictError>, _>(|p| {
+						Ok(PathBeneath::new(PathFd::new(p)?, access_read))
+					}),
+			)?
+			.add_rules(
+				self.ro_dirs
+					.as_slice()
+					.iter()
+					.map::<Result<_, RestrictError>, _>(|p| {
+						Ok(PathBeneath::new(PathFd::new(p)?, access_dir_with_rm))
+					}),
+			)?
+			.set_compatibility(self.compat_level)
+			.set_no_new_privs(true)
+			.restrict_self()?)
```

Comment:
I'd suggest to add a global (boolean/Option) variable + Mutex to track if the landlock restrictions have already been applied to the current process and fail gracefully if so.